SubZero/src/model/interface/HwInterface.h

```diff
@@ -1,27 +1,156 @@
-/*
+/**
  * HwInterface.h
  *
- *  Created on: Jan 17, 2015
- *      Author: ahsueh1996
+ * Created on: Jan 17, 2015
+ * Author: ahsueh1996
  */
 
 #ifndef HWINTERFACE_H_
 #define HWINTERFACE_H_
 
+#include "Data.h"
+// Data is the generic datatype defined in Data.h
+
+
 /**
- * This abstract class is responsible for all the io interfacing with hardware.
+ * HwInterface is an abstract class responsible for all the io interfacing with hardware.
  */
 class HwInterface {
+
+private:
+
+	/**
+	 * Pointer to memory allocated for storing raw (but
+	 * decoded) data from hardware.
+	 */
+	Data** decodedBuffer[];
+
+	/**
+	 * Allocate only enough memory to keep a max of
+	 * this number of hardware inputs; discard all after that.
+	 */
+	int bufferSize;
+
+	/**
+	 * Pull data from hardware at this frequency.
+	 */
+	int pullFrequency;
+
+	/**
+	 * Specifies the encoding, decoding policy to use.
+	 */
+	int policy;
+	/*
+	 * alternatively, we can just implement encode(), decode() ... etc
+	 * differently for each hardwares when we branch out
+	 * from our abstract class
+	 */
+
+	/*
+	 * ID of hardware this interface interacts with
+	 * hardware can be accessed (somehow) via this ID
+	 */
+	int hardwareID;
+
+	/**
+	 * Private helper function for send().
+	 * Encode the message at pointer Data src (source)
+	 * @param src, des
+	 */
+	virtual void encode(Data* src);
+
+	/**
+	 * this class will be automatically pulling and managing
+	 * the pulling process privately within the interface at pullFrequency
+	 * using the following functions
+	 */
+
+	/**
+	 * Pull raw data from hardware
+	 */
+	virtual void pull();
+
+	/**
+	 * Decode the data at src.
+   	 */
+	virtual Data* decode();
+
+protected:
+
+	/**
+	 * Delete buffer from startIdx to endIdx
+	 * @param startIdx, endIdx
+	 */
+	virtual void deleteFromBuffer(int startIdx, int endIdx);
+
+
+	/**
+	  * Store decoded data at src to buffer
+	  * @param src
+	  */
+	virtual void storeToBuffer(Data* src);
+
 public:
-	HwInterface();
+
 	/**
-	 * Constructor
+	 * Return the most recent buffer data
+	 *
 	 */
-	virtual ~HwInterface();
+	virtual void getDataFromBuffer();
```

Comment:
Oh.... hmm.... didn't think about that. Right now it looks like the children are implementing it. If we change it to virtual Data* getDataFromBuffer() then the parent can implement it. Yeah, maybe that's better. Thanks. :+1: